### PR TITLE
add benchmark & speed up implementation

### PR DESCRIPTION
**Pull request by yoshuawuyts**
## Changes
- adds a benchmark dir
- speeds up `pull.js` implementation by a reasonable amount

Thanks!

### CODE REVIEW

**Comment by dominictarr** [comment]
the benchmark is called node-streams.js but I don't see any node streams in it?
I'm gonna merge this, but I don't really approve of this benchmark, because it doesn't tell us very much.
hmm, probably benchmarks should be their own thing (module) so that it's easy to test different versions of the code.

**Comment by dominictarr** [comment]
oh, I see you didn't run the tests... );

File: pull.js [hunk 1]
```diff
@@ -1,50 +1,28 @@
...
-
-module.exports = function pull () {
-  var args = [].slice.call(arguments)
-
-  if(isReader(args[0]))
+module.exports = function pull (a) {
+  if (typeof a === 'function' && a.length === 1) {
     return function (read) {
-      args.unshift(read)
+      var args = [].slice.call(arguments)
```
Comment (dominictarr) [inline pull.js]: this is scope bug. if it's inside the function args will be `[read]`

**Comment by yoshuawuyts** [comment]
very fair; I'll remove the bench file. Also: hmmm, let's see if I can fix that test

**Comment by yoshuawuyts** [comment]
ah cool, you fixed the bug :tada:

**Comment by dominictarr** [comment]
Nevermind, I fixed it in 3.2.1
I also added some more ways of creating a pipeline to the benchmark, so at now it shows that `pull(source, pull(through, sink))` is slower (after I fixed the bug)

**Comment by dominictarr** [comment]
```
pull3*100000: 1444.374ms
pull_compose*100000: 2255.037ms
pull_chain*100000: 1536.539ms
```
I wonder if there is a faster way to do that though?

**Comment by dominictarr** [comment]
this `fastbench` thing is pretty simple, but I wish that it showed, say ops/second, and the relative perf of the different runs... and maybe some statistics?

**Comment by yoshuawuyts** [comment]
I believe that's what https://github.com/bestiejs/benchmark.js/ tries to do; it's def less simple though

**Pull request merged**